Request Controllers #8

seblaifa3315 opened the pull request:
### What this PR does (required):
Created an endpoint for the following related to requests:
- GET : list of all the requests for the logged in user (as an owner and as a sitter)
- POST : Create a new request
- UPDATE : Update accepted and declined properties of the request

### Any information needed to test this feature (required):
- For this ticket, I used the Request model created in the "feature_request_model" branch who hasn't been review yet.
Let me know if you want me to change the model and I will do the fixes in the controllers accordingly.

### Any issues with the current functionality (optional):
- for now the GET request list of the requests to the logged in user no matter if it is as an owner or a sitter. Maybe in the future we can create a GET that lists only the request as an owner and another GET that lists only the request as a sitter.

ezeilo-su commented on server/controllers/request.js:
```diff
@@ -0,0 +1,68 @@
+
+exports.createRequest = async(req, res, next) => {   
+    const { user_id, sitter_id } = req.body;
+    // Check if owner exists
```
Comment is not needed!

ezeilo-su commented on server/controllers/request.js:
```diff
@@ -0,0 +1,68 @@
+        throw new Error("Can't find any pet owner with this id");
+    }
+
+    //check if sitter exists
```
Remove this comment

ezeilo-su commented on server/controllers/request.js:
```diff
@@ -0,0 +1,68 @@
+        throw new Error("Can't find any petsitter with this id")
+    }
+
+    //create a Request Document and send it if data is valid
```
Also remove this one

ezeilo-su commented on server/controllers/request.js:
```diff
@@ -0,0 +1,68 @@
+exports.getRequests = async(req, res, next) => {
+    const { userId }  = req.body;
+
+    //check if the user exists
```
Please remove comment

ezeilo-su commented on server/controllers/request.js:
```diff
@@ -0,0 +1,68 @@
+exports.updateRequest = async(req, res, next) => {
+    const { userId, requestId, accepted, declined }  = req.body;
+
+    //check if the user exists
```
Get rid of this comment

ezeilo-su commented on server/controllers/request.js:
```diff
@@ -0,0 +1,68 @@
+        throw new Error("Can't find any user with this id");
+    };
+
+    //find a request document with the requestId && the userId and update accepted and declined properties
```
Also remove this one

ezeilo-su commented on server/models/Request.js:
```diff
@@ -0,0 +1,56 @@
+
+const requestSchema = new mongoose.Schema(
+    {
+        user_id: {
```
Use `camelCase` in naming your variables

ezeilo-su commented on server/models/Request.js:
```diff
@@ -0,0 +1,56 @@
+            required: true,
+            ref: 'User'
+        },
+        sitter_id: {
```
`camelCase` this one too

ezeilo-su reviewed (changes requested):
Hey @seblaifa3315, good work with this one. But please address those comments and also use the right linters setup to format your code.
Happy coding!

seblaifa3315 commented:
- I addressed the above comments
- I still didn't use the right linters setup to format my code... still have to learn how to do it, sorry!!!

seblaifa3315 commented:
- I  addressed the comments above
- I made some modifications in the logic of the controllers
- I changed the routes of request (added routes)
- I still didn't use the right linters setup to format the code... I still have to learn how to do that, sorry!!!

rajivtitus commented on server/app.js:
```diff
@@ -45,6 +46,7 @@ app.use((req, res, next) => {
 app.use("/auth", authRouter);
 app.use("/users", userRouter);
 app.use("/profile", profileRouter);
+app.use("/request", requestRouter);
```
```suggestion
app.use("/requests", requestRouter);
```
Let's rename this route as per REST API naming convention

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,99 @@
+exports.createRequest = async (req, res, next) => {
+    const { ownerId, sitterId } = req.body;
+
+    const ownerExists = await User.findOne({ ownerId });
```
💯

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,99 @@
+const Request = require("../models/Request");
+
+exports.createRequest = async (req, res, next) => {
+    const { ownerId, sitterId } = req.body;
```
Let's also include the start and end dates within the request body here

rajivtitus replied:
Optional: We can also validate the request body params before passing them to mongoose queries

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,99 @@
+const Profile = require("../models/Profile");
+const Request = require("../models/Request");
+
+exports.createRequest = async (req, res, next) => {
```
Since we're not using the `asyncHandler` here, let's wrap our async operations with a try/catch block 

Ditto where applicable

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,99 @@
+};
+
+exports.getRequests = async (req, res, next) => {
+    const { userId } = req.body;
```
It is not recommended to pass anything in the request body of a GET request. We can use the request object to obtain the logged in user's id 

Eg: `req.user.id`

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,99 @@
+};
+
+exports.getSitterRequests = async (req, res, next) => {
+    const { id } = req.params;
```
Please refer to comment above for obtaining the logged in user's id

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,99 @@
+    }
+};
+
+exports.getOwnerRequests = async (req, res, next) => {
```
This controller and the one above it can be combined into one. Once we have the user's id, we can find the relevant profile and conditionally querying the db based on the account type

rajivtitus commented on server/models/Request.js:
```diff
@@ -0,0 +1,70 @@
+const mongoose = require("mongoose");
+
+const requestSchema = new mongoose.Schema(
```
Please rebase/adjust the base of this PR to only reflect the changes pertaining to this ticket

rajivtitus commented on server/routes/request.js:
```diff
@@ -0,0 +1,12 @@
+const { createRequest, getRequests, getSitterRequests, getOwnerRequests, updateRequest } = require("../controllers/request");
+
+router.route("/").post(protect, createRequest);
+router.route("/").get(protect, getRequests);
```
The three GET routes can be consolidate into one route :)

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,99 @@
+        throw new Error("Can't find any user with this id");
+    }
+
+    const requestUpdated = await Request.findOneAndUpdate({_id: requestId }, { accepted: accepted, declined: declined }, { new: true });
```
We need to ensure that only the sitter who owns the request should be able to edit it

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,99 @@
+};
+
+exports.updateRequest = async (req, res, next) => {
+    const { sitterId, requestId, accepted, declined, avatar } = req.body;
```
```suggestion
    const { accepted, declined } = req.body;
```
I don't think we won't to be able to change who the owner/sitter is after a request has been made

rajivtitus reviewed (commented):
Good start Sebastien! Please address the comments for the Request model and update this branch. Once done, you can start working on the comments in this PR 👍🏽

rajivtitus commented:
> I made the changes according to the comments above... except 2 comments, because I'm not sure I understood exactly what they mean:
> 
> 1. "Please rebase/adjust the base of this PR to only reflect the changes pertaining to this ticket":
>    I adjusted the Request Model to be able to test my controllers... was it wrong?
> 2. "Optional: We can also validate the request body params before passing them to mongoose queries":
>    Should i check id the dates, animalType etc are valid there? Or should it be done by the request Model validations?

Sure, please see below for additional context:

1. When we work on a feature which is dependent on a previous branch, we branch off of it - which is completely normal.  We also need to make sure that we:
  - [Changing the base branch of a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request) **OR**
  - [Rebase the PR](https://github.com/openedx/edx-platform/wiki/How-to-Rebase-a-Pull-Request)
2. It is a good practice to validate the request body parameters (startDate, endDate, sitterId et cetera) 
  - We can do this within the controller using an `if` clause **OR**
  - Use a middleware like express validator: https://www.freecodecamp.org/news/how-to-make-input-validation-simple-and-clean-in-your-express-js-app-ea9b5ff5a8a7/

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,83 @@
+
+exports.createRequest = async (req, res) => {
+    const { ownerId, sitterId, start, end, animalType, usefulInfo } = req.body;
+    console.log(new Date());
```
Please remove log to the console

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,83 @@
+
+        res.status(200).json(request);
+    } catch (error) {
+        res.status(404).json({ message: error.message });
```
```suggestion
        res.status(500).json({ message: error.message });
```
A `500` seems more appropriate here

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,83 @@
+        if (!sitterExists) return res.status(404).json({ message: "Sitter doesn't exist" });
+
+        const request = await Request.create(req.body);
+        if (!request) return res.status(404).json({ message: "unable to create a new request" });
```
```suggestion
```
I don't think we'll require this line if a mongoose create query fails (the error will be handled by our catch block)

rajivtitus commented on server/routes/request.js:
```diff
@@ -0,0 +1,10 @@
+
+router.route("/").post(protect, createRequest);
+router.route("/").get(protect, getRequests);
+router.route("/").put(protect, updateRequest);
```
```suggestion
router.route("/:requestId").put(protect, updateRequest);
```
We'll need a path param to identify the specific resource we wish to identify/modify

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,83 @@
+
+        if(typeOfAccount === 'pet_sitter') {
+            const requests = await Request.find({ sitterId: req.user.id }).lean();
+            if (!requests) return res.status(404).json({ message: "Something went wrong!!!" });
```
```suggestion
```
A mongoose `find()` query returns an empty list, we can return this to the FE to have a consistent response format.  Also, the above `if` clause will always evaluate to false as an empty list is a truthy value.

Ditto where applicable

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,83 @@
+        }
+        if(typeOfAccount === 'pet_owner') {
+            const requests = await Request.find({ ownerId: req.user.id }).lean();
+            if (!requests) return res.status(404).json({ message: "Something went wrong!!!" });
```
Same as comment above

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,83 @@
+        }
+            
+    } catch (error) {
+        res.status(404).json({ message: error.message });
```
```suggestion
        res.status(500).json({ message: error.message });
```

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,83 @@
+        if (!userExists) return res.status(404).json({ message: "User doesn't exist" });
+
+        const theRequest = await Request.findOne({ _id: requestId });
+        if(theRequest.sitterId.valueOf() !== req.user.id) return res.status(404).json({ message: "This sitter doesn't own this request" });
```
💯

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,83 @@
+
+        res.status(200).json(requestUpdated);
+    } catch (error) {
+        res.status(404).json({ message: error.message });
```
```suggestion
        res.status(500).json({ message: error.message });
```

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,83 @@
+        const theRequest = await Request.findOne({ _id: requestId });
+        if(theRequest.sitterId.valueOf() !== req.user.id) return res.status(404).json({ message: "This sitter doesn't own this request" });
+
+        const requestUpdated = await Request.findOneAndUpdate({ _id: requestId }, { status: newStatus }, { new: true });
```
```suggestion
    theRequest.status = newStatus;
    await theRequest.save()
```
We don't need to query the db again, we can edit the existing request object and save it

rajivtitus reviewed (commented):
Good work with the change Sebastien! A few more comments based on the new code, please let me know if you have any questions

seblaifa3315 commented:
I made the changes according to the comments above.
I just left the optional one ("validate the request body params before passing them to mongoose queries")... I think I need to study more about it before I implement it.

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,80 @@
+
+        const request = await Request.create(req.body);
+
+        res.status(200).json(request);
```
```suggestion
        res.status(201).json(request);
```

rajivtitus reviewed (approved):
LGTM! Great work with those changes 👍🏽
